From fe395e7f4be3a8435a0140273856e1ab1628081f Mon Sep 17 00:00:00 2001
From: Sergey Senozhatsky <senozhatsky@chromium.org>
Date: Sat, 13 Feb 2021 13:17:28 +0900
Subject: [PATCH 9/9] videobuf2: handle non-contiguous DMA allocations

This adds support for new DMA API. New DMA API requires allocators
to have two execution branches: one for the current API, and one for
the new DMA API, which depends on
	a) dma_can_alloc_noncontiguous() support
	b) user-space non-coherent allocation request

Signed-off-by: Sergey Senozhatsky <senozhatsky@chromium.org>
[hch: untested conversion to the ne API]
Signed-off-by: Christoph Hellwig <hch@lst.de>
---
 .../common/videobuf2/videobuf2-dma-contig.c   | 134 +++++++++++++++---
 1 file changed, 111 insertions(+), 23 deletions(-)

diff --git a/drivers/media/common/videobuf2/videobuf2-dma-contig.c b/drivers/media/common/videobuf2/videobuf2-dma-contig.c
index 9ad6d890b85a..78f5b4a6d91a 100644
--- a/drivers/media/common/videobuf2/videobuf2-dma-contig.c
+++ b/drivers/media/common/videobuf2/videobuf2-dma-contig.c
@@ -27,6 +27,7 @@ struct vb2_dc_buf {
 	void				*vaddr;
 	unsigned long			size;
 	void				*cookie;
+	struct sg_table			*nc_sgt;
 	dma_addr_t			dma_addr;
 	unsigned long			attrs;
 	enum dma_data_direction		dma_dir;
@@ -43,6 +44,11 @@ struct vb2_dc_buf {
 	struct vb2_buffer		*vb;
 };
 
+static bool vb2_is_coherent(struct vb2_dc_buf *buf)
+{
+	return !buf->nc_sgt;
+}
+
 /*********************************************/
 /*        scatterlist table functions        */
 /*********************************************/
@@ -77,12 +83,18 @@ static void *vb2_dc_cookie(void *buf_priv)
 static void *vb2_dc_vaddr(void *buf_priv)
 {
 	struct vb2_dc_buf *buf = buf_priv;
-	struct dma_buf_map map;
-	int ret;
 
 	if (!buf->vaddr && buf->db_attach) {
-		ret = dma_buf_vmap(buf->db_attach->dmabuf, &map);
-		buf->vaddr = ret ? NULL : map.vaddr;
+		if (vb2_is_coherent(buf)) {
+			struct dma_buf_map map;
+
+			if (!dma_buf_vmap(buf->db_attach->dmabuf, &map))
+				buf->vaddr = map.vaddr;
+		} else {
+			buf->vaddr = dma_vmap_noncontiguous(buf->dev,
+							    buf->size,
+							    buf->nc_sgt);
+		}
 	}
 
 	return buf->vaddr;
@@ -100,9 +112,22 @@ static void vb2_dc_prepare(void *buf_priv)
 	struct vb2_dc_buf *buf = buf_priv;
 	struct sg_table *sgt = buf->dma_sgt;
 
+	/* This takes care of DMABUF and user-enforced cache sync hint */
 	if (buf->vb->no_cache_sync_on_prepare)
 		return;
 
+	/* We can test ->nc_sgt only for MMAP buffers */
+	if (buf->vb->memory == V4L2_MEMORY_MMAP) {
+		if (vb2_is_coherent(buf))
+			return;
+		sgt = buf->nc_sgt;
+	}
+
+
+	/*
+	 * sgt points to two different tables: ->nc_sgt for V4L2_MEMORY_MMAP
+	 * buffers and to ->dma_sgt otherwise
+	 */
 	if (!sgt)
 		return;
 
@@ -114,9 +139,21 @@ static void vb2_dc_finish(void *buf_priv)
 	struct vb2_dc_buf *buf = buf_priv;
 	struct sg_table *sgt = buf->dma_sgt;
 
+	/* This takes care of DMABUF and user-enforced cache sync hint */
 	if (buf->vb->no_cache_sync_on_finish)
 		return;
 
+	/* We can test ->nc_sgt only for MMAP buffers */
+	if (buf->vb->memory == V4L2_MEMORY_MMAP) {
+		if (vb2_is_coherent(buf))
+			return;
+		sgt = buf->nc_sgt;
+	}
+
+	/*
+	 * sgt points to two different tables: ->nc_sgt for V4L2_MEMORY_MMAP
+	 * buffers and to ->dma_sgt otherwise
+	 */
 	if (!sgt)
 		return;
 
@@ -127,6 +164,21 @@ static void vb2_dc_finish(void *buf_priv)
 /*        callbacks for MMAP buffers         */
 /*********************************************/
 
+static void __vb2_dc_put(struct vb2_dc_buf *buf)
+{
+	struct vb2_queue *q = buf->vb->vb2_queue;
+
+	if (vb2_is_coherent(buf)) {
+		dma_free_attrs(buf->dev, buf->size, buf->cookie,
+			       buf->dma_addr, q->dma_attrs);
+		return;
+	}
+
+	if (buf->vaddr)
+		dma_vunmap_noncontiguous(buf->dev, buf->vaddr);
+	dma_free_noncontiguous(buf->dev, buf->size, buf->nc_sgt, buf->dma_addr);
+}
+
 static void vb2_dc_put(void *buf_priv)
 {
 	struct vb2_dc_buf *buf = buf_priv;
@@ -138,12 +190,47 @@ static void vb2_dc_put(void *buf_priv)
 		sg_free_table(buf->sgt_base);
 		kfree(buf->sgt_base);
 	}
-	dma_free_attrs(buf->dev, buf->size, buf->cookie, buf->dma_addr,
-		       buf->attrs);
+	__vb2_dc_put(buf);
 	put_device(buf->dev);
 	kfree(buf);
 }
 
+static int __vb2_dc_alloc(struct vb2_dc_buf *buf)
+{
+	struct vb2_queue *q = buf->vb->vb2_queue;
+
+	if (!q->non_consistent_mem) {
+		buf->cookie = dma_alloc_attrs(buf->dev, buf->size,
+					      &buf->dma_addr,
+					      GFP_KERNEL | q->gfp_flags,
+					      q->dma_attrs);
+		if (!buf->cookie)
+			return -ENOMEM;
+		if ((q->dma_attrs & DMA_ATTR_NO_KERNEL_MAPPING) == 0)
+			buf->vaddr = buf->cookie;
+		return 0;
+	}
+
+	buf->nc_sgt = dma_alloc_noncontiguous(buf->dev, buf->size,
+					      buf->dma_dir,
+					      GFP_KERNEL | q->gfp_flags);
+	if (!buf->nc_sgt)
+		return -ENOMEM;
+
+	if (q->dma_attrs & DMA_ATTR_NO_KERNEL_MAPPING)
+		return 0;
+
+	buf->vaddr = dma_vmap_noncontiguous(buf->dev, buf->size, buf->nc_sgt);
+	if (!buf->vaddr)
+		goto err;
+	return 0;
+
+err:
+	dma_free_noncontiguous(buf->dev, buf->size, buf->nc_sgt, buf->dma_addr);
+	buf->nc_sgt = NULL;
+	return -ENOMEM;
+}
+
 static void *vb2_dc_alloc(struct vb2_buffer *vb,
 			  struct device *dev,
 			  unsigned long size)
@@ -158,27 +245,22 @@ static void *vb2_dc_alloc(struct vb2_buffer *vb,
 		return ERR_PTR(-ENOMEM);
 
 	buf->attrs = vb->vb2_queue->dma_attrs;
-	buf->cookie = dma_alloc_attrs(dev, size, &buf->dma_addr,
-				      GFP_KERNEL | vb->vb2_queue->gfp_flags,
-				      buf->attrs);
-	if (!buf->cookie) {
-		dev_err(dev, "dma_alloc_coherent of size %ld failed\n", size);
-		kfree(buf);
-		return ERR_PTR(-ENOMEM);
-	}
-
-	if ((buf->attrs & DMA_ATTR_NO_KERNEL_MAPPING) == 0)
-		buf->vaddr = buf->cookie;
+	buf->dma_dir = vb->vb2_queue->dma_dir;
+	buf->vb = vb;
 
+	buf->size = size;
 	/* Prevent the device from being released while the buffer is used */
 	buf->dev = get_device(dev);
-	buf->size = size;
-	buf->dma_dir = vb->vb2_queue->dma_dir;
+
+	if (__vb2_dc_alloc(buf)) {
+		dev_err(dev, "dma alloc of size %ld failed\n", size);
+		kfree(buf);
+		return ERR_PTR(-ENOMEM);
+	}
 
 	buf->handler.refcount = &buf->refcount;
 	buf->handler.put = vb2_dc_put;
 	buf->handler.arg = buf;
-	buf->vb = vb;
 
 	refcount_set(&buf->refcount, 1);
 
@@ -195,9 +277,12 @@ static int vb2_dc_mmap(void *buf_priv, struct vm_area_struct *vma)
 		return -EINVAL;
 	}
 
-	ret = dma_mmap_attrs(buf->dev, vma, buf->cookie,
-		buf->dma_addr, buf->size, buf->attrs);
-
+	if (vb2_is_coherent(buf))
+		ret = dma_mmap_attrs(buf->dev, vma, buf->cookie, buf->dma_addr,
+				     buf->size, buf->attrs);
+	else
+		ret = dma_mmap_noncontiguous(buf->dev, vma, buf->size,
+					     buf->nc_sgt);
 	if (ret) {
 		pr_err("Remapping memory failed, error: %d\n", ret);
 		return ret;
@@ -389,6 +474,9 @@ static struct sg_table *vb2_dc_get_base_sgt(struct vb2_dc_buf *buf)
 	int ret;
 	struct sg_table *sgt;
 
+	if (!vb2_is_coherent(buf))
+		return buf->nc_sgt;
+
 	sgt = kmalloc(sizeof(*sgt), GFP_KERNEL);
 	if (!sgt) {
 		dev_err(buf->dev, "failed to alloc sg table\n");
-- 
2.30.1

